### PR TITLE
Database reconnect utility (Resolves #233)

### LINKED ISSUE #233: Retry queries until successful
**Describe the bug**
We need to retry all queries in case the db connection dies.

**Affected Component(s)**
- [x] Back-End (Database, Microservices, Containers, etc)
- [ ] Front-End (Flask, API, etc)
- [ ] Docs
- [ ] Build System

**To Reproduce**
1. Start application and monitor module
2. Restart postgres database
3. Listener and Database modules exceptions

### PR DESCRIPTION
**Pull request by slowr**
### Description of PR

A automated way to reconnect to the database if the connection is dropped and also retry queries.

What component(s) does this PR affect?

- [x] Back-End (Database, Microservices, Containers, etc)
- [ ] Front-End (Flask, API, etc)
- [ ] Docs
- [ ] Build System

Does the PR require changes on other components? If yes, please mark the components:

- [ ] Back-End (Database, Microservices, Containers, etc)
- [ ] Front-End (Flask, API, etc)
- [ ] Docs
- [ ] Build System

### Related Issue

Resolves #233 

### Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update
- [ ] None of the above

### Checklist:

- [x] I have read the **[contributing guide](https://github.com/FORTH-ICS-INSPIRE/artemis/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [ ] This change requires a change in the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: backend/core/utils/tool.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                retry_counter = 0
+                self._connection.set_session(
+                    autocommit=self.autocommit, readonly=self.readonly
+                )
+                return self._connection
+            except psycopg2.OperationalError as error:
+                if not self.reconnect or retry_counter >= LIMIT_RETRIES:
+                    raise error
+                retry_counter += 1
+                print(
```
Comment (vkotronis) [inline backend/core/utils/tool.py]: print or log.error? Same for the rest of the prints.

File: backend/core/utils/tool.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            retry_counter += 1
+            print("got error {}. retrying {}".format(str(error).strip(), retry_counter))
+            time.sleep(1)
+            self.reset()
+            self.execute(query, vals, retry_counter)
+        except (Exception, psycopg2.Error) as error:
+            if not self.readonly:
+                self._connection.rollback()
+            raise error
+        else:
```
Comment (vkotronis) [inline backend/core/utils/tool.py]: where does this else correspond?

File: backend/core/utils/tool.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                raise error
+
+    def cursor(self):
+        if not self._cursor or self._cursor.closed:
+            if not self._connection:
+                self.connect()
+            self._cursor = self._connection.cursor()
+            return self._cursor
+
+    def execute(self, query, vals=None, retry_counter=0):
```
Comment (vkotronis) [inline backend/core/utils/tool.py]: maybe we can have a single function to reuse code? execute, execute_batch and execute_values are more than 90% the same. The difference is in the main call and in the extra page_size param for the values, right?

File: backend/core/utils/tool.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            retry_counter += 1
+            print("got error {}. retrying {}".format(str(error).strip(), retry_counter))
+            time.sleep(1)
+            self.reset()
+            self.execute_batch(query, vals, retry_counter)
+        except (Exception, psycopg2.Error) as error:
+            if not self.readonly:
+                self._connection.rollback()
+            raise error
+        else:
```
Comment (vkotronis) [inline backend/core/utils/tool.py]: same for this else.

File: backend/core/utils/tool.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            retry_counter += 1
+            print("got error {}. retrying {}".format(str(error).strip(), retry_counter))
+            time.sleep(1)
+            self.reset()
+            self.execute_values(query, vals, page_size, retry_counter)
+        except (Exception, psycopg2.Error) as error:
+            if not self.readonly:
+                self._connection.rollback()
+            raise error
+        if not self.readonly:
```
Comment (vkotronis) [inline backend/core/utils/tool.py]: this cannot be readonly this is why there is no related code here right?

File: backend/core/utils/tool.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                    port=self.port,
+                    database=self.database,
+                    connect_timeout=3,
+                )
+                retry_counter = 0
+                self._connection.set_session(
+                    autocommit=self.autocommit, readonly=self.readonly
+                )
+                return self._connection
+            except psycopg2.OperationalError as error:
```
Comment (vkotronis) [inline backend/core/utils/tool.py]: what about psycopg2.DatabaseError?

Comment (slowr) [reply backend/core/utils/tool.py]: don't think it's needed for this part. it's when trying to execute mostly
http://initd.org/psycopg/docs/errors.html

File: backend/core/utils/tool.py (hunk 1 above)
Comment (slowr) [reply backend/core/utils/tool.py]: good point forgot the logger

File: backend/core/utils/tool.py (hunk 3 above)
Comment (slowr) [reply backend/core/utils/tool.py]: it was more clear for me doing the migration. we can optimize the code afterwards

File: backend/core/utils/tool.py (hunk 2 above)
Comment (slowr) [reply backend/core/utils/tool.py]: if there is no exception and it's a write sql command commit the changes

File: backend/core/utils/tool.py (hunk 4 above)
Comment (slowr) [reply backend/core/utils/tool.py]: same as before

File: backend/core/utils/tool.py (hunk 5 above)
Comment (slowr) [reply backend/core/utils/tool.py]: readonly checks are for commit and rollback. in our cases execute_values only write

File: backend/core/utils/tool.py (hunk 3 above)
Comment (vkotronis) [reply backend/core/utils/tool.py]: Ok. Resolving.

**Comment by vkotronis** [comment]
@slowr pending error for the PR listener impl (not to forget):
```
 database - 2019-09-21 18:15:26,822 - ERROR @ _retrieve_most_recent_config_hash: failed to retrieved most recent config hash in db
backend           | Traceback (most recent call last):
backend           |   File "database.py", line 1482, in _retrieve_most_recent_config_hash
backend           |     )[0]
backend           | IndexError: list index out of range
```
Could you also sum up in a comment the issue with the DB duplicates that comes up with the new approach for db writes?

File: backend/core/database.py [hunk 7]
```diff
@@ -1500,27 +1471,18 @@ def _update_bgp_updates(self):
...
                     # execute serial execute_batch query
                     query = "UPDATE bgp_updates SET handled=true, hijack_key=array_distinct(hijack_key || array[%s]) WHERE bgp_updates.key=%s"
-                    with get_wo_cursor(self.wo_conn) as db_cur:
-                        psycopg2.extras.execute_batch(
-                            db_cur,
-                            query,
-                            list(update_bgp_entries_serial),
-                            page_size=1000,
-                        )
+                    self.wo_db.execute_values(
```
Comment (vkotronis) [inline backend/core/database.py]: this should be execute_batch

File: backend/core/database.py [hunk 8]
```diff
@@ -17,14 +16,15 @@
...
+from utils import DB_PORT
+from utils import DB_USER
 from utils import flatten
-from utils import get_db_conn
 from utils import get_hash
 from utils import get_ip_version
 from utils import get_logger
-from utils import get_ro_cursor
-from utils import get_wo_cursor
-from utils import hijack_log_field_formatter
```
Comment (vkotronis) [inline backend/core/database.py]: why was this deleted?

File: backend/core/database.py [hunk 9]
```diff
@@ -77,28 +78,23 @@ def __init__(self):
...
         try:
             with Connection(RABBITMQ_URI) as connection:
-                self.worker = self.Worker(connection, ro_conn, wo_conn)
+                self.worker = self.Worker(connection)
                 self.worker.run()
         except Exception:
             log.exception("exception")
         finally:
             log.info("stopped")
-            ro_conn.close()
```
Comment (vkotronis) [inline backend/core/database.py]: does it automatically close now?

**Comment by slowr** [comment]
:running: **Benchmark Results** :running:
- BGP Updates (updates/sec): **644**
- Hijacks Detected (hijacks/sec): **25**

File: backend/core/database.py (hunk 8 above)
Comment (slowr) [reply backend/core/database.py]: by mistake, added back

File: backend/core/database.py (hunk 9 above)
Comment (slowr) [reply backend/core/database.py]: yes

**Comment by slowr** [comment]
:running: **Benchmark Results** :running:
- BGP Updates (updates/sec): **704**
- Hijacks Detected (hijacks/sec): **27**

**Comment by slowr** [comment]
:running: **Benchmark Results** :running:
- BGP Updates (updates/sec): **630**
- Hijacks Detected (hijacks/sec): **23**

**Comment by slowr** [comment]
:running: **Benchmark Results** :running:
- BGP Updates (updates/sec): **636**
- Hijacks Detected (hijacks/sec): **27**

**Comment by slowr** [comment]
:running: **Benchmark Results** :running:
- BGP Updates (updates/sec): **725**
- Hijacks Detected (hijacks/sec): **28**

**Comment by slowr** [comment]
:running: **Benchmark Results** :running:
- BGP Updates (updates/sec): **702**
- Hijacks Detected (hijacks/sec): **26**

File: backend/core/utils/tool.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python
+import time
+
+import psycopg2.extras
+from utils import get_logger
+
+ISOLEVEL = psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT
+
+LIMIT_RETRIES = 5
```
Comment (vkotronis) [inline backend/core/utils/tool.py]: is this enough by the  way? Maybe we should have a large value?

**Comment by vkotronis** [comment]
LGTM!
